### PR TITLE
[inductor] Reduce duplication of reduction combine functions

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102737
* #102486
* #102736
* __->__ #99661

Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that inplace reduction operations are now written as
an out-of-place operation and an assignment, e.g. instead if
```cpp
omp_out += omp_in
```
we generate
```cpp
omp_out = omp_out + omp_in
```

Which is a purely cosmetic change

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1327,19 +1327,8 @@ def final_argreduce(buffer, result_var, value, index):
...
-                elif reduction_type == "xor_sum":
-                    updated = f"{accumulator} ^ {value}"
-                else:
-                    raise NotImplementedError(f"reduction_type {reduction_type}")
-
+                combine_fn = ir.get_reduction_combine_fn(reduction_type, src_dtype)
+                updated = combine_fn(accumulator, value)
                 self.compute.writeline(
                     f"{accumulator} = tl.where({cond}, {updated}, {accumulator})"
                 )
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1334]: Idea, but not for this PR. If we could know that `value` has just been loaded, we could elide this `tl.where`, as in those cases, the `other=` kwarg within `tl.load` already sets those extra values to the correct identity element.

I don't think we have a nice way of doing this now though, as the CSE vars carry close to no information.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1327,19 +1327,8 @@ def final_argreduce(buffer, result_var, value, index):
...
-                elif reduction_type == "sum":
-                    updated = f"{accumulator} + {value}"
-                elif reduction_type == "prod":
-                    updated = f"{accumulator} * {value}"
-                elif reduction_type == "xor_sum":
-                    updated = f"{accumulator} ^ {value}"
-                else:
-                    raise NotImplementedError(f"reduction_type {reduction_type}")
-
+                combine_fn = ir.get_reduction_combine_fn(reduction_type, src_dtype)
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1330]: `arg{max,min}` do not go through this path, and rather re-implement everything. Any chance we can merge those two paths as well?

**Review by lezcano — COMMENTED** [review]
To generate C++ code that is a bit less cryptic, it'd be good to have an `inplace=False` kwarg on `reduction_combine_vec`.

**Review by lezcano — COMMENTED** [review]
To generate C++ code that is a bit less cryptic, it'd be good to have an `inplace=False` kwarg on `reduction_combine_vec`.

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (peterbell10) [reply torch/_inductor/codegen/triton.py:1330]: tbh I think `triton_helpers.minimum_with_index` is a lot more readable than generating the whole combine function.

Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1330]: That's right. This was more of a

Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1330]: That's right. The only thing is that, for that path, we still have two implementations of the same thing, but I guess that's alright.

**Review by lezcano — APPROVED** [review]

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1330]: It'd be good to have a note asking people to keep these two in sync if they modify one or the other. No need to do it in this PR tho.

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -1403,28 +1396,20 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
...
-                reduce_all_body += (
-                    f"return {vec_ns}::{reduce_map[reduction_type]}(x, y);"
-                )
-            reduce_all_body += "}"
+            reduce_all_body = (
+                "{ return " + reduction_combine_vec(reduction_type, "x", "y") + "; }"
+            )
             vec_reduce_all_func = f"{vec_ns}::vec_reduce_all<{DTYPE_TO_CPP[dtype]}>"
-            next_value = f"{vec_reduce_all_func}([]({vec}& x, {vec}&y) {reduce_all_body}, {tmpvar_vec})"
+            next_value = f"{vec_reduce_all_func}([]({vec}& x, {vec}& y) {reduce_all_body}, {tmpvar_vec})"
```
Comment (lezcano) [inline torch/_inductor/codegen/cpp.py:1408]: Nit. The previous change of passing these by value was potentially beneficial. As these variables fit in registers, they are better passed by value. Perhaps the compiler is able to optimise out this lambda though.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**